File: Tree/Tree.py
from collections import deque

class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
    
    # Вставить значение
    def insert(self, value):
        if self.root is None:
            self.root = TreeNode(value)
        else:
            self._insert_recursive(self.root, value)
    
    def _insert_recursive(self, node, value):
        if value < node.value:
            if node.left is None:
                node.left = TreeNode(value)
            else:
                self._insert_recursive(node.left, value)
        elif value > node.value:
            if node.right is None:
                node.right = TreeNode(value)
            else:
                self._insert_recursive(node.right, value)
    
    # Поиск значения
    def search(self, value):
        return self._search_recursive(self.root, value)
    
    def _search_recursive(self, node, value):
        if node is None:
            return False
        
        if value == node.value:
            return True
        elif value < node.value:
            return self._search_recursive(node.left, value)
        else:
            return self._search_recursive(node.right, value)
    
    # Средний обход (Inorder)
    def inorder_traversal(self, node=None, result=None):
        if result is None:
            result = []
        if node is None:
            node = self.root
        
        if node:
            self.inorder_traversal(node.left, result)
            result.append(node.value)
            self.inorder_traversal(node.right, result)
        
        return result
    
    # Прямой обход (Preorder)
    def preorder_traversal(self, node=None, result=None):
        if result is None:
            result = []
        if node is None:
            node = self.root
        
        if node:
            result.append(node.value)
            self.preorder_traversal(node.left, result)
            self.preorder_traversal(node.right, result)
        
        return result
    
    # Обратный обход (Postorder)
    def postorder_traversal(self, node=None, result=None):
        if result is None:
            result = []
        if node is None:
            node = self.root
        
        if node:
            self.postorder_traversal(node.left, result)
            self.postorder_traversal(node.right, result)
            result.append(node.value)
        
        return result
    
    # Обход в ширину (Level-order)
    def level_order_traversal(self):
        if not self.root:
            return []
        
        result = []
        queue = deque([self.root])
        
        while queue:
            node = queue.popleft()
            result.append(node.value)
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        return result
    
    # Найти минимальное значение
    def find_min(self, node=None):
        if node is None:
            node = self.root
        
        while node.left:
            node = node.left
        
        return node.value
    
    # Найти максимальное значение
    def find_max(self, node=None):
        if node is None:
            node = self.root
        
        while node.right:
            node = node.right
        
        return node.value

# Пример использования
if __name__ == "__main__":
    bst = BinarySearchTree()
    
    bst.insert(50)
    bst.insert(30)
    bst.insert(70)
    bst.insert(20)
    bst.insert(40)
    bst.insert(60)
    bst.insert(80)
    
    print('Средний обход (Inorder):', bst.inorder_traversal())
    print('Прямой обход (Preorder):', bst.preorder_traversal())
    print('Обратный обход (Postorder):', bst.postorder_traversal())
    print('Обход в ширину (Level-order):', bst.level_order_traversal())
    
    print('Поиск 40:', bst.search(40))  # True
    print('Поиск 100:', bst.search(100))  # False
    
    print('Минимум:', bst.find_min())  # 20
    print('Максимум:', bst.find_max())  # 80

File: Tree/Tree.js
// Класс для узла бинарного дерева
class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

// Класс для бинарного дерева поиска
class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  // Вставить значение
  insert(value) {
    const newNode = new TreeNode(value);
    
    if (this.root === null) {
      this.root = newNode;
      return this;
    }
    
    let current = this.root;
    while (true) {
      if (value === current.value) return undefined; // Дубликаты не допускаются
      
      if (value < current.value) {
        if (current.left === null) {
          current.left = newNode;
          return this;
        }
        current = current.left;
      } else {
        if (current.right === null) {
          current.right = newNode;
          return this;
        }
        current = current.right;
      }
    }
  }

  // Поиск значения
  search(value) {
    let current = this.root;
    
    while (current) {
      if (value === current.value) return true;
      if (value < current.value) {
        current = current.left;
      } else {
        current = current.right;
      }
    }
    
    return false;
  }

  // Средний обход (Inorder) - рекурсивно
  inorderTraversal(node = this.root, result = []) {
    if (node !== null) {
      this.inorderTraversal(node.left, result);
      result.push(node.value);
      this.inorderTraversal(node.right, result);
    }
    return result;
  }

  // Прямой обход (Preorder)
  preorderTraversal(node = this.root, result = []) {
    if (node !== null) {
      result.push(node.value);
      this.preorderTraversal(node.left, result);
      this.preorderTraversal(node.right, result);
    }
    return result;
  }

  // Обратный обход (Postorder)
  postorderTraversal(node = this.root, result = []) {
    if (node !== null) {
      this.postorderTraversal(node.left, result);
      this.postorderTraversal(node.right, result);
      result.push(node.value);
    }
    return result;
  }

  // Обход в ширину (Level-order)
  levelOrderTraversal() {
    if (!this.root) return [];
    
    const result = [];
    const queue = [this.root];
    
    while (queue.length > 0) {
      const node = queue.shift();
      result.push(node.value);
      
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    
    return result;
  }

  // Найти минимальное значение
  findMin(node = this.root) {
    while (node.left !== null) {
      node = node.left;
    }
    return node.value;
  }

  // Найти максимальное значение
  findMax(node = this.root) {
    while (node.right !== null) {
      node = node.right;
    }
    return node.value;
  }
}

// Пример использования
const bst = new BinarySearchTree();

bst.insert(50);
bst.insert(30);
bst.insert(70);
bst.insert(20);
bst.insert(40);
bst.insert(60);
bst.insert(80);

console.log('Средний обход (Inorder):', bst.inorderTraversal());
console.log('Прямой обход (Preorder):', bst.preorderTraversal());
console.log('Обратный обход (Postorder):', bst.postorderTraversal());
console.log('Обход в ширину (Level-order):', bst.levelOrderTraversal());

console.log('Поиск 40:', bst.search(40)); // true
console.log('Поиск 100:', bst.search(100)); // false

console.log('Минимум:', bst.findMin()); // 20
console.log('Максимум:', bst.findMax()); // 80
